app/service/data_svc.py

```diff
@@ -19,6 +19,25 @@
 from app.utility.base_service import BaseService
 
 Adjustment = namedtuple('Adjustment', 'ability_id trait value offset')
+history = defaultdict(set)
+
+
+def track_recursion(f):
```

Comment:
this strikes me as being better suited as a base_service function, not tied to the data_svc directly. thoughts?

Reply:
I had that thought as well, though as you notice it is outside the data_svc class, and most of our imports are the class from the module not the whole module.  But I agree it could be useful outside of here.